lib/Runtime/Library/Utf8String.h

```diff
@@ -0,0 +1,142 @@
+
+            charcount_t utf16Length = 0;
+            utf8::DecodeOptions opts = utf8::DecodeOptions::doDefault;
+            LPCUTF8 buf = reinterpret_cast<LPCUTF8>(buffer);
+            LPCUTF8 end = buf + utf8Length;
+            while (buf < end)
+            {
+                if ((*buf & 0x80) == 0)
+                {
+                    // Single byte character
+                    utf16Length++;
+                    buf++;
+                }
+                else
+                {
+                    // Decode a single utf16 character, and increment the pointer
+                    // to the start of the next character.
+                    char16 c1 = *buf;
+                    ++buf;
+                    utf8::DecodeTail(c1, buf, end, opts);
+                    utf16Length++;
+                }
+            }
+
+            this->SetLength(utf16Length);
+            this->SetBuffer(nullptr);
+        }
+
+
+        size_t Utf8Length() const
+        {
+            return this->utf8String->length;
+        }
+
+        const char* Utf8Buffer() const
+        {
+            return this->utf8String->buffer;
+        }
+
+        const char16* GetSz() override
+        {
+            if (this->IsFinalized())
+            {
+                return this->UnsafeGetBuffer();
+            }
+
+            // TODO: This is currently wrong in the presence of unmatched surrogate pairs.
+            // They will be converted to a 3 byte replacement character in utf8, and then
+            // converted back into a 1 utf16 character 0xfffd representation of that, not the original
+            // utf16 surrogate pair character
+            char16* buffer = RecyclerNewArrayLeaf(this->GetRecycler(), char16, this->GetLength() + 1);
+
+            // Fetching the char* from the Field(char*) first so we can then cast to LPCUTF8
+            const char* bufferStart = this->utf8String->buffer;
+            LPCUTF8 start = reinterpret_cast<LPCUTF8>(bufferStart);
+            size_t decodeLength = utf8::DecodeUnitsIntoAndNullTerminateNoAdvance(buffer, start, start + this->utf8String->length);
+
+            Assert(decodeLength == this->GetLength());
+
+            buffer[this->GetLength()] = 0;
+
+            this->SetBuffer(buffer);
+            return this->UnsafeGetBuffer();
+        }
+
+        static bool Is(RecyclableObject* obj)
+        {
+            return VirtualTableInfo<Js::Utf8String>::HasVirtualTable(obj);
+        }
+
+        static Utf8String* From(RecyclableObject* obj)
+        {
+            if (Utf8String::Is(obj))
+            {
+                return static_cast<Utf8String*>(obj);
+            }
+
+            return nullptr;
+        }
+
+        template <typename StringType>
+        static Utf8String * ConvertString(StringType * originalString, _In_reads_(utf8Length) char* buffer, size_t utf8Length)
+        {
+            VirtualTableInfo<Utf8String>::SetVirtualTable(originalString);
```

Comment:
I kind of expected SetVirtualTable to be a template that static-asserted that the source type was big enough for the destination type, but I don't see any such assertion. Could you please add one? Otherwise this is concerning because LiteralString is not big enough to convert successfully.